smartmeter demo (CHERIoT device bits)

nwf opened the pull request.

nwf committed:
First draft of smartmeter demo CHERIoT bits

nwf committed:
Allow speeding up simulated time

This should allow making the demo more interesting

rmn30 committed:
Change xmake default paths to suit cheriot-demos repo.

Also change the name of CHERIOT_SDK env variable to CHERIOT_RTOS_SDK to avoid confusion with --sdk option to xmake.

rmn30 committed:
Add smart_meter demo to CI.

Also set default board for smart_meter demo to sonata-1.1.

nwf committed:
broker configuration via xmake

nwf committed:
fix client IDs

We self-check for validity only if debugging is turned on, and how odd
that '-' isn't in the spec's permitted set (3.1.3.1, MQTT-3.1.3-5).

nwf committed:
grid: report some data, not just a Tick

nwf committed:
Very, very initial web frontend goo

vmurali commented:
Looks like I cannot push to this branch. Here's a slightly different design to highlight the security guarantees of CHERIoT against a system that has no compartmentalization:

1. The grid controller

   1. Receives scheduled outages (array of <start time, duration>)
   2. Recieves request for corrective actions i.e. load shedding, load increase (array of <start time, duration, maximum load value, minimum load value>)

2. The provider

   1. Receives hourly rate information for the next day (<date, array of 24 prices>); the old rates are used till a new rate is received for a day
   2. Sends accumulated consumption information hourly
   3. Optional: receives a directive to turn off the energy for various reasons (payment not made, hourly consumption information isa not received for a long time)

3. The sensor: provides a passive interface for reading the following values
   1. Instantaneous total load consumption in the house
   2. Optional: Instantaneous energy stored in each battery in the house (array of <battery index, energy>)

4. The house
   1. Updates/accumulates the hourly comsumption information based on energy consumption sensor data (reads sensor data frequently)
   2. Provides a passive interface to specify the load to be consumed for an array of load consumers (array of <consumer index, load> updated frequently)
   3. Optional: Provides a passive interface to specify the status of charge/discharge for an array of batteries (array of <battery index, CHARGE/DISCHARGE> updated frequently)

5. The user
   1. Provides a passive interface to specify the policy (using a Javascript bytecode) to drive the load consumers (and optionally battery) based on grid request, provider rate (and optionally battery energy)
   2. Calls the house interface to specify load consumer values and battery charge/discharge status based on the current policy

The grid controller and provider each run a thread for receiving data in the form of MQTT messages, and the provider runs a thread for sending accumulated consumption information.
The provider has web interface to show the current accumulated (monthly) charges.
The sensor can optionally have a web interface to show the current load consumption and batteries' levels. The house runs a thread the updates the hourly consumption information.
The user provides a web interface to input the javascript program to specify policy.
The user runs a thread to update the loads to be consumed by various consumers and optionally the charge/discharge status for the batteries.
Optional: The grid and provider has a web interface to submit new rates, requests, outages etc.

The javascript VM has FFI functions to access 
  1. the outage information, () -> (Start Time, Duration)
  2. grid load requests, () -> (Start Time, Duration, Max Load Value, Min Load Value)
  3. provider rate information for the current day, Hour -> Rate
  4. sensor load consumption, () -> Current load
  5. battery level, Battery Index -> Current Energy Stored
  6. set load consumptions, Consumer Index -> Load -> ()
  7. set battery charge/discharge, Battery Index -> CHARGE/DISCHARGE -> ()

On a system with no security features, a user can write a javascript program that abuses the set load consumptions or set battery charge/discharge functions to specify an arbitrary 32-bit index to update any value in the memory.
In particular, the hourly consumption information can be overwritten to 0 to make the provider think that the total energy consumption is 0 throughout.

With CHERIoT, each of the above components can be specified as a separate compartment and as a result, the user compartment that has the javascript VM cannot access/update arbitrary memory using the index value in the set load consumptions/set battery charge/discharge functions.

Note that the grid controller is optional for this security feature demonstration and can potentially be removed to simplify the design.

nwf committed:
WIP: monolithic image as well

nwf committed:
Incremental progress on user frontend

nwf committed:
Monolithic build

Works after https://github.com/CHERIoT-Platform/cheriot-rtos/pull/461

nwf committed:
sample_policy: use spaces, not tabs

nwf committed:
Add "revert to last submitted" button to editor

nwf committed:
Further set the stage for badness

On monolithic builds, move the sensor data and the userJS data snapshot
objects into (the one compartment's) globals.  This should make it
possible, in a monolithic build, for a cap held in the JS machine to be
abused to write to the sensor data.

nwf committed:
Further demolish monolith security

Place sensitive and non-sensitive structures not just in the same .data
segment but as subobjects of the same struct.  This subverts
CHERIoT-LLVM's understanding of properly bounding address-taken globals
and permits subobject bounds violtion to model code running on a
pre-CHERI machine.

rmn30 committed:
Add 'unique-id' xmake option for setting meter ID at compile time.

Defaults to 'random' which causes the device to generate a random ID on boot.

rmn30 committed:
Update network stack to include fix for MQTT packet bounds.

nwf committed:
frontend: more functionality in grid view

nwf committed:
frontend: rename directory holding on-device JS

nwf committed:
Expand CHERIoT component README a bit

nwf committed:
frontend: user-editor rollup config file

nwf committed:
frontend democonfig: add default meter ID

rmn30 committed:
Fix comment in ffigen.js and regenerate userffi.h.

rmn30 committed:
Configure uart1 and add js ffi for writing to it.

This means we can now have user js talk to esp32 over uart.

rmn30 committed:
Add some smart home python files.

nwf committed:
frontend: first stab at provider UI

Update CHERIoT readme and dox in light of changes here, too

nwf committed:
Add proposal for house protocol

rmn30 committed:
Add RPi demo js config and add an option to allow self-signed cert.

rmn30 committed:
Add notes on setting up RPi server for smartmeter demo.

rmn30 committed:
Smart meter: further work on house python.

nwf committed:
frontends: add consumption report graph

rmn30 committed:
initial attempt at reading powerSamples from house. Had to disable monolith build due to odd link error.

nwf committed:
device: use explicit timeouts in MQTT connect

nwf committed:
device: sensor error handler only in non-monolith

rmn30 committed:
Re-enable the monolith build.

https://github.com/CHERIoT-Platform/cheriot-rtos/pull/472 worked
around the linker bug that prevented this so bump cheriot-rtos version
to include this and fixup the xmake. We have to drop the error handler
from sensor compartment as it conflicts in the monolith build. I only
added it to help debug something anyway.

nwf committed:
device: prepare for coarse and fine sensor data

nwf committed:
device sensor: coarse data, too

nwf committed:
frontend: allow relative paths for broker certs

Have the server interpret them relative to the static resource
directory, as they would be by the clients (served out of the static
directory).

nwf committed:
frontend: user editor dropdown for provided examples

rmn30 committed:
Smart meter demo: latest version of house python with updated lcd display

nwf committed:
device: report user JS crashes to grid

nwf committed:
grid frontend: report crash counts

nwf committed:
treewide: overhaul js

nwf committed:
clang-format

nwf committed:
device: optional fake sensor reports

nwf committed:
device provider: publish samples

nwf committed:
frontends: consumption updates might be negative

nwf committed:
Add omitted file

nwf committed:
Further tweaking of policy files

nwf committed:
device: tweak sensor coarsening

- Give symbolic name to the coarsening depth
- Fix a rather embarassing bug

nwf committed:
Attack demo

nwf committed:
Tweak default policy

And make it really the default JS baked in.  While here, make that
process easier and document it.

nwf committed:
Add crash example

nwf committed:
device: refactor monolith build; separate user

This lets the JS crash and get forcibly unwound back to a calling
compartment rather than to thread exit.

nwf committed:
Revise monolith again

This partially reverts ca4e80255e49a6514ce6a990e4130c9ae475fa2b and puts
the merged data object back into the monolith compartment's .data.  The
separated user compartment can get it via a cross-compartment call
during initialization.  This means that it can be accessed via the CGP
exposed to JS, too, in addition to the array pointer exposed via the
FFI.

nwf committed:
frontends: arrays are 0 indexed, d'oh.

While here, sort the consumption arrays so we know we're dropping
prefixes, just in case timestamps wiggle a bit.

nwf committed:
Emit newlines when sending to house

nwf committed:
device sensor: attempt to make reads more robust

- Move to IRQ-based reads, rather than burning the core
- Bump the sensor thread priority over the workers'
- Prevent std::string from consuming unbounded heap
- Have std::string perform allocation up front and Debug::Assert success

nwf committed:
device sensor: reuse std::string for lines

We're seeing somewhat mysterious allocation failures, so try to reduce
the number of allocations we make?

nwf committed:
device js: add export to inject network faults

nwf committed:
smartmeter device user: fix multiwater use

multiwaiter_wait() overwrites the value fields of its EventWaiterSource
structures, so cache the values passed in.  Update the cache and copy
from it back to the wait events each loop iteration.

nwf committed:
smartmeter: coarsen samples over many minutes

nwf committed:
smartmeter: deeper housekeeping stacks

We were running out of stack deep in the network code, spinning in a
tight loop of failing to allocate a network buffer.  Don't do that.

rmn30 committed:
Add more notes and config files for server config on Mac

rmn30 committed:
Add new cheriot.demo cert as the old one expired.

Also move it to /static dir as that's where it needs to be.

rmn30 committed:
Add index.html with iframes for the three views.

Adjust graph width properties to make this work.

nwf committed:
smartmeter: lower SNTP timeouts

The FreeRTOS-Plus-TCP network stack drops with success UDP datagrams it
cannot immediately send, so our first attempt at this is all but
guaranteed to fail at present.  Don't wait quite so long for that to
happen, and don't wait so long between attempts.  Continue to allow the
2nd and subsequent attempts to take longer.
